pkg/documentlog/store/sql.go

```diff
@@ -0,0 +1,263 @@
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	})
+	return consistencyHash, err
+}
+
+func (s sqlDocumentStore) Get(hash model.Hash) (*StoredDocument, error) {
+	document := sqlDocument{}
+	err := s.db.
+		Table(documentTable).
+		Select(documentColumns).
+		Where("hash = ?", hash.String()).
+		First(&document).
+		Error
+	if errors.Is(err, gorm.ErrRecordNotFound) {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	result := document.storedDocument()
+	return &result, nil
+}
+
+func (s sqlDocumentStore) GetByConsistencyHash(hash model.Hash) (*StoredDocument, error) {
+	document := sqlDocument{}
+	err := s.db.
+		Table(documentTable).
+		Select(documentColumns).
+		Where("consistency_hash = ?", hash.String()).
+		First(&document).
+		Error
+	if errors.Is(err, gorm.ErrRecordNotFound) {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	result := document.storedDocument()
+	return &result, nil
+}
+
+func (s sqlDocumentStore) GetAll() ([]StoredDocument, error) {
+	documents := make([]sqlDocument, 0)
+	err := s.db.Table(documentTable).Select(documentColumns).Find(&documents).Error
+	if err != nil {
+		return nil, err
+	}
+	result := make([]StoredDocument, len(documents))
+	for i, document := range documents {
+		result[i] = document.storedDocument()
+	}
+	return result, nil
+}
+
+func (s sqlDocumentStore) WriteContents(hash model.Hash, contents io.Reader) error {
+	// TODO: I'd rather stream the contents to the blob, instead of reading it as byte buffer which might be problematic
+	// with large documents
+	buf := new(bytes.Buffer)
+	if _, err := buf.ReadFrom(contents); err != nil {
+		return err
+	}
+	return s.db.Table(documentTable).Where("hash = ?", hash.String()).UpdateColumn("contents", buf.Bytes()).Error
+}
+
+func (s sqlDocumentStore) ReadContents(hash model.Hash) (io.ReadCloser, error) {
+	document := sqlDocument{}
+	err := s.db.
+		Table(documentTable).
+		Select("contents").
+		Where("hash = ?", hash.String()).
+		First(&document).Error
+	if err != nil {
+		return nil, err
+	}
+	return &NoopCloser{Reader: bytes.NewReader(document.Contents)}, nil
+}
+
+func (s sqlDocumentStore) LastConsistencyHash() model.Hash {
+	values := make([]interface{}, 1)
+	query := s.db.
+		Table(documentTable).
+		Select("consistency_hash").
+		Where("timestamp = ?", s.db.Table(documentTable).Select("MAX(timestamp)").SubQuery()).
+		Limit(1).
+		Pluck("consistency_hash", &values)
+	if query.Error != nil {
+		logging.Log().Errorf("Unable to determine last consistency hash: %v", query.Error)
+		return model.EmptyHash()
+	}
+	// We map to interface{} to account for NULL value
+	if len(values) == 0 {
+		return model.EmptyHash()
+	}
+	hash, _ := model.ParseHash(values[0].(string))
+	return hash
+}
+
+func (s sqlDocumentStore) ContentsSize() int {
+	values := make([]interface{}, 1)
+	query := s.db.Table(documentTable).Pluck("SUM(LENGTH(contents))", &values)
```

Comment:
I've looked into this for SQLite which provides ANALYZE for database stats, but this gets slow for large databases. File size can be misleading since it doesn't account space that's not reclaimed yet (using compacting/gaeage collection) and only works when the actual database is on the same machine as nuts-go is.

Maybe introduce a stats-table which tracks the value? Which can then be updated by 1. the application when writing document contents or 2. a trigger